crates/parse/src/parser/mod.rs

```diff
@@ -54,6 +59,13 @@ pub struct Parser<'sess, 'ast> {
     in_yul: bool,
     /// Whether the parser is currently parsing a contract block.
     in_contract: bool,
+
+    /// Current recursion depth for expression parsing via `parse_expr`.
+    expr_recursion_depth: usize,
```

Comment:
I think we should also have a single counter that we call with `with_recursion_limit(f: impl FnOnce(&mut Self) -> PResult<T>) -> PResult<T>`, it can also take in a string for the context like "expression"